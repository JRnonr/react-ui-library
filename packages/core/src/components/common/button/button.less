/* 基础按钮样式 */
.velvet-btn {
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: var(--spacing-sm);
  height: 32px;
  padding: 4px 15px;
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-md);
  background-color: var(--bg-color);
  color: var(--text-color);
  font-size: var(--font-size-sm);
  font-weight: var(--font-weight-normal);
  line-height: 1.5715;
  white-space: nowrap;
  text-align: center;
  cursor: pointer;
  user-select: none;
  touch-action: manipulation;
  transition: all var(--transition-normal);
  outline: none;
  box-sizing: border-box;
  margin: 0;

  /* 悬停态（非禁用时） */
  &:hover:not(:disabled):not(.velvet-btn--disabled) {
    color: var(--primary-color);
    border-color: var(--primary-color);
    background-color: var(--primary-color-light);
  }

  /* 激活态（非禁用时） */
  &:active:not(:disabled):not(.velvet-btn--disabled) {
    color: var(--primary-color-active);
    border-color: var(--primary-color-active);
    background-color: var(--primary-color-light);
  }

  /* 键盘聚焦可见态 */
  &:focus-visible {
    outline: 2px solid var(--primary-color-light);
    outline-offset: 1px;
  }

  /* 主要按钮 */
  &.velvet-btn--primary {
    color: var(--text-color-inverse);
    background-color: var(--primary-color);
    border-color: var(--primary-color);

    &:hover:not(:disabled):not(.velvet-btn--disabled) {
      background-color: var(--primary-color-hover);
      border-color: var(--primary-color-hover);
      color: var(--text-color-inverse);
    }

    &:active:not(:disabled):not(.velvet-btn--disabled) {
      background-color: var(--primary-color-active);
      border-color: var(--primary-color-active);
    }
  }

  /* 虚线按钮 */
  &.velvet-btn--dashed {
    border-style: dashed;
  }

  /* 文本按钮 */
  &.velvet-btn--text {
    color: var(--text-color);
    background-color: transparent;
    border-color: transparent;
    padding: 0;
    height: auto;

    &:hover:not(:disabled):not(.velvet-btn--disabled) {
      color: var(--text-color-secondary);
      background-color: var(--bg-color-light);
    }

    &:active:not(:disabled):not(.velvet-btn--disabled) {
      color: var(--text-color);
      background-color: var(--bg-color-dark);
    }
  }

  /* 禁用状态（属性和类名双重支持） */
  &[disabled],
  &[aria-disabled="true"],
  &.velvet-btn--disabled {
    cursor: not-allowed;
    pointer-events: none;
    border-color: var(--border-color);
    color: var(--text-color-disabled);
    background-color: var(--bg-color-light);
    box-shadow: none;

    /* 禁用时禁止任何交互样式 */
    &:hover,
    &:focus,
    &:active {
      cursor: not-allowed;
      pointer-events: none;
      border-color: var(--border-color);
      color: var(--text-color-disabled);
      background-color: var(--bg-color-light);
      box-shadow: none;
    }

    /* 链接按钮禁用状态 - 无边框无背景但保持下划线 */
    &.velvet-btn--link {
      border: none !important;
      border-color: transparent !important;
      background-color: transparent !important;
      box-shadow: none !important;
      text-decoration: underline;
      
      &:hover,
      &:focus,
      &:active {
        border: none !important;
        border-color: transparent !important;
        background-color: transparent !important;
        box-shadow: none !important;
        text-decoration: underline;
      }
    }

    /* 文本按钮禁用状态 - 无边框无背景 */
    &.velvet-btn--text {
      border-color: transparent;
      background-color: transparent;
      
      &:hover,
      &:focus,
      &:active {
        border-color: transparent;
        background-color: transparent;
      }
    }
  }
}

/* 链接按钮 - 使用更高特异性确保样式生效 */
button.velvet-btn.velvet-btn--link,
.velvet-btn.velvet-btn--link {
  color: var(--primary-color) !important;
  background-color: transparent !important;
  border: 0 !important;
  border-width: 0 !important;
  border-style: none !important;
  border-color: transparent !important;
  box-shadow: none !important;
  padding: 0 !important;
  height: auto !important;
  text-decoration: underline !important;

  &:hover:not(:disabled):not(.velvet-btn--disabled),
  &:focus:not(:disabled):not(.velvet-btn--disabled),
  &:active:not(:disabled):not(.velvet-btn--disabled) {
    background-color: transparent !important;
    border: 0 !important;
    border-width: 0 !important;
    border-style: none !important;
    border-color: transparent !important;
    box-shadow: none !important;
    text-decoration: underline !important;
  }

  &:hover:not(:disabled):not(.velvet-btn--disabled) {
    color: var(--primary-color-hover) !important;
  }

  &:active:not(:disabled):not(.velvet-btn--disabled) {
    color: var(--primary-color-active) !important;
  }

  &:focus-visible {
    outline: 2px solid var(--primary-color-light) !important;
    outline-offset: 1px !important;
    border: 0 !important;
    box-shadow: none !important;
  }

  /* 禁用状态 */
  &[disabled],
  &[aria-disabled="true"],
  &.velvet-btn--disabled {
    border: 0 !important;
    border-width: 0 !important;
    border-style: none !important;
    border-color: transparent !important;
    background-color: transparent !important;
    box-shadow: none !important;
    text-decoration: underline !important;
    color: var(--text-color-disabled) !important;
    
    &:hover,
    &:focus,
    &:active {
      border: 0 !important;
      border-width: 0 !important;
      border-style: none !important;
      border-color: transparent !important;
      background-color: transparent !important;
      box-shadow: none !important;
      text-decoration: underline !important;
      color: var(--text-color-disabled) !important;
    }
  }
}
