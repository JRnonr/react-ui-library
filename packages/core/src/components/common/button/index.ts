// 从当前目录下的 button.tsx 文件中导出 Button 组件和 ButtonProps 类型
// 这样外部只需要 import { Button } from 'components/common/button'
// 而不必直接写 import { Button } from 'components/common/button/button';

export { Button } from './button';
export type { ButtonProps, ButtonType } from './button'; 