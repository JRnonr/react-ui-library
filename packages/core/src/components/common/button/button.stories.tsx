import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './button';

// Storybook 的元信息，告诉 Storybook 这个组件的标题、组件、参数等
const meta: Meta<typeof Button> = {
  title: 'Components/Button', // Storybook 左侧菜单的名字
  component: Button, // 要展示的组件
  parameters: {
    layout: 'centered', // 整个故事居中显示
  },
  argTypes: {
    disabled: {
      control: { type: 'boolean' }, // 给 Storybook 加个勾选框控制 disabled 属性
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;  //定义了一个 Story 类型，方便后面写不同场景

// 基础按钮 这是第一个场景 名字叫 Basic（Storybook 里会显示这个名字）。
export const Basic: Story = {
  render: () => ( //render 决定怎么画这个按钮，这里是最普通的按钮。
    <div style={{ 
      display: 'flex', 
      gap: '12px', 
      alignItems: 'center',
      flexWrap: 'wrap',
      padding: '20px',
      backgroundColor: '#f5f5f5',
      borderRadius: '8px'
    }}>
      <Button>基础按钮</Button>
    </div>
  ),
};

// 禁用状态 这是第二个场景 名字叫 Disabled（Storybook 里会显示这个名字）。
export const Disabled: Story = {
  render: () => (
    <div style={{ 
      display: 'flex', 
      gap: '12px', 
      alignItems: 'center',
      flexWrap: 'wrap',
      padding: '20px',
      backgroundColor: '#f5f5f5',
      borderRadius: '8px'
    }}>
      <Button disabled>禁用按钮</Button>
    </div>
  ),
}; 