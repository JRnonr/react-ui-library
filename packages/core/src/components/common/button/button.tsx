import React, { forwardRef, ButtonHTMLAttributes, ReactNode } from 'react';
import classNames from 'classnames';
import '../../../styles/design-system.css';
import './button.less';

export type ButtonType = 'default' | 'primary' | 'dashed' | 'link' | 'text';

export interface ButtonProps extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'> {
  /** 按钮类型 */
  type?: ButtonType;
  /** 是否禁用按钮 */
  disabled?: boolean;
  /** 点击事件的回调函数 */
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  /** 按钮内部的子节点 */
  children?: ReactNode;
  /** 原生按钮类型 */
  htmlType?: 'submit' | 'reset' | 'button';
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      type = 'default',
      disabled = false,
      onClick,
      children,
      style,
      htmlType = 'button',
      ...rest
    },
    ref
  ) => {
    const buttonClasses = classNames(
      'velvet-btn',
      {
        [`velvet-btn--${type}`]: type && type !== 'default',
        'velvet-btn--disabled': disabled
      },
      className
    );

    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
      if (disabled) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      onClick?.(event);
    };

    return (
      <button
        ref={ref}
        className={buttonClasses}
        disabled={disabled}
        onClick={handleClick}
        style={style}
        type={htmlType}
        aria-disabled={disabled}
        {...rest}
      >
        {children}
      </button>
    );
  }
);

Button.displayName = 'Button';
