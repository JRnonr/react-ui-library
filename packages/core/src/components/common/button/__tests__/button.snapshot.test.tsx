import React from 'react';
import { render } from '@testing-library/react';
import { Button } from '../button';

describe('Button 快照测试', () => {
  it('默认按钮应该匹配快照', () => {
    const { container } = render(<Button>测试按钮</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('primary类型按钮应该匹配快照', () => {
    const { container } = render(<Button type="primary">主要按钮</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('dashed类型按钮应该匹配快照', () => {
    const { container } = render(<Button type="dashed">虚线按钮</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('link类型按钮应该匹配快照', () => {
    const { container } = render(<Button type="link">链接按钮</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('text类型按钮应该匹配快照', () => {
    const { container } = render(<Button type="text">文本按钮</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('禁用状态按钮应该匹配快照', () => {
    const { container } = render(<Button disabled>禁用按钮</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('带自定义样式的按钮应该匹配快照', () => {
    const { container } = render(
      <Button style={{ backgroundColor: 'red', color: 'white' }}>
        自定义样式按钮
      </Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('带自定义类名的按钮应该匹配快照', () => {
    const { container } = render(
      <Button className="custom-button">自定义类名按钮</Button>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
}); 