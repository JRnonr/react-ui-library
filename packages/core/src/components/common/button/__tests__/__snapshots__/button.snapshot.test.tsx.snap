// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 快照测试 dashed类型按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn velvet-btn--dashed"
  type="button"
>
  虚线按钮
</button>
`;

exports[`Button 快照测试 link类型按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn velvet-btn--link"
  type="button"
>
  链接按钮
</button>
`;

exports[`Button 快照测试 primary类型按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn velvet-btn--primary"
  type="button"
>
  主要按钮
</button>
`;

exports[`Button 快照测试 text类型按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn velvet-btn--text"
  type="button"
>
  文本按钮
</button>
`;

exports[`Button 快照测试 带自定义样式的按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn"
  style="background-color: red; color: white;"
  type="button"
>
  自定义样式按钮
</button>
`;

exports[`Button 快照测试 带自定义类名的按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn custom-button"
  type="button"
>
  自定义类名按钮
</button>
`;

exports[`Button 快照测试 禁用状态按钮应该匹配快照 1`] = `
<button
  aria-disabled="true"
  class="velvet-btn velvet-btn--disabled"
  disabled=""
  type="button"
>
  禁用按钮
</button>
`;

exports[`Button 快照测试 默认按钮应该匹配快照 1`] = `
<button
  aria-disabled="false"
  class="velvet-btn"
  type="button"
>
  测试按钮
</button>
`;
